README.md
# NAME

IIIF - IIIF Image API implementation

[![MetaCPAN Release](https://badge.fury.io/pl/IIIF.svg)](https://metacpan.org/release/IIIF)
[![Linux Build Status](https://travis-ci.com/nichtich/IIIF.svg?branch=master)](https://travis-ci.com/nichtich/IIIF)
[![Windows Build Status](https://ci.appveyor.com/api/projects/status/dko0d7647jvfgu8w?svg=true)](https://ci.appveyor.com/project/nichtich/iiif)
[![Coverage Status](https://coveralls.io/repos/nichtich/IIIF/badge.svg)](https://coveralls.io/r/nichtich/IIIF)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/IIIF.png)](http://cpants.cpanauthors.org/dist/IIIF)

# DESCRIPTION

Package IIIF provides an implementation of [IIIF ImageAPI](https://iiif.io/api/image/3.0/)
based on the [ImageMagick](https://www.imagemagick.org/) command line application.

# FEATURES

- Full [IIIF ImageAPI 3.0](https://iiif.io/api/image/3.0/) level 2 compliance
- Support abbreviated requests (e.g. `300,200` to select size, `90/gray` to
select rotation and quality...).
- Web service ([IIIF::ImageAPI](https://metacpan.org/pod/IIIF::ImageAPI)) and command line client ([i3f](https://metacpan.org/pod/i3f))
- Heavily tested
    - passing the [IIIF Image API Validator](https://iiif.io/api/image/validator/)
    - works with ImageMagick 6 (tested on Ubuntu Linux) and ImageMagick 7 (tested on Windows)
    - 100% [test coverage](https://coveralls.io/github/nichtich/IIIF) on statement
    level, (>90% on branch level and >70% on condition level).

# INSTALLATION

See also ["REQUIREMENTS" in IIIF::Magick](https://metacpan.org/pod/IIIF::Magick#REQUIREMENTS) for additional installation for optional
features.

# UNIX

Most Unixes include system Perl by default. You should also install ImageMagick and
[cpanminus](https://metacpan.org/pod/App::cpanminus#INSTALLATION). For instance at
Ubuntu Linux:

    sudo apt-get install imagemagick cpanminus

To speed up installation of Perl dependencies of this package, optionally:

    sudo apt-get install libplack-perl libplack-middleware-crossorigin-perl

And for optional support of WebP format:

    sudo apt-get install webp libwebp-dev

Then install IIIF with Perl package manager:

    cpanm IIIF

## WINDOWS

Install ImageMagick and Perl, for instance with `Chocolatey|https://chocolatey.org`:

    choco install imagemagick.tool
    choco install strawberryperl

Then install IIIF with Perl package manager:

    cpanm IIIF

# MODULES

- [IIIF::Request](https://metacpan.org/pod/IIIF::Request)

    parse and express an IIIF Image API request build of region, size, rotation, quality, and format

- [IIIF::Magick](https://metacpan.org/pod/IIIF::Magick)

    get image information and convert images as specified with IIIF Image API request using ImageMagick

- [IIIF::ImageAPI](https://metacpan.org/pod/IIIF::ImageAPI)

    provide a [Plack](https://metacpan.org/pod/Plack) web service to access images via IIIF Image API

# SCRIPTS

This module provides the command line script [i3f](https://metacpan.org/pod/i3f) to apply IIIF Image API requests without a web service.

# SEE ALSO

- [https://github.com/IIIF/awesome-iiif](https://github.com/IIIF/awesome-iiif)
- [Image::Magick](https://metacpan.org/pod/Image::Magick)
- [Plack::App::ImageMagick](https://metacpan.org/pod/Plack::App::ImageMagick)

# LICENSE

Copyright (C) Jakob Voß.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Jakob Voß <voss@gbv.de>


Changes
Revision history for Perl extension IIIF

{{$NEXT}}

0.07 2019-12-09T12:21:26Z
    - add canonicalLinkHeader (#16)
    - support maxWidth and maxHeight (to some degree, see #6)

0.06 2019-10-24T10:28:48Z
    - Extend documentation
    - Propagate features in JSON response
    - Fix WebP and JPEG 2000 MIME types
    - Fix protocol in info.json

0.05 2019-10-11T09:42:41Z
    - major cleanup

0.04 2019-10-10T14:58:56Z
    - add option magick_args

0.03 2019-10-10T12:50:26Z
    - implement canonical requests
    - add base URI parameter
    - improve test coverage

0.02 2019-10-09T15:24:04Z
    - extend documentation and error handling
    - implement size parameter

0.01 2019-10-09T11:40:02Z
    - initial version



lib/IIIF.pm
package IIIF;
use 5.014001;

our $VERSION = "0.06";

1;
__END__

=encoding utf-8

=head1 NAME

IIIF - IIIF Image API implementation

=begin markdown 

[![MetaCPAN Release](https://badge.fury.io/pl/IIIF.svg)](https://metacpan.org/release/IIIF)
[![Linux Build Status](https://travis-ci.com/nichtich/IIIF.svg?branch=master)](https://travis-ci.com/nichtich/IIIF)
[![Windows Build Status](https://ci.appveyor.com/api/projects/status/dko0d7647jvfgu8w?svg=true)](https://ci.appveyor.com/project/nichtich/iiif)
[![Coverage Status](https://coveralls.io/repos/nichtich/IIIF/badge.svg)](https://coveralls.io/r/nichtich/IIIF)
[![Kwalitee Score](http://cpants.cpanauthors.org/dist/IIIF.png)](http://cpants.cpanauthors.org/dist/IIIF)

=end markdown

=head1 DESCRIPTION

Package IIIF provides an implementation of L<IIIF ImageAPI|https://iiif.io/api/image/3.0/>
based on the L<ImageMagick|https://www.imagemagick.org/> command line application: Requests
to get a specific segment of an image are mapped to command line arguments of ImageMagick
to perform the requested segment extraction. See L<i3f/EXAMPLES> for examples.

=head1 FEATURES

=over

=item

Full L<IIIF ImageAPI 3.0|https://iiif.io/api/image/3.0/> level 2 compliance

=item

Support abbreviated requests (e.g. C<300,200> to select size, C<90/gray> to
select rotation and quality...).

=item

Web service (L<IIIF::ImageAPI>) and command line client (L<i3f>)

=item

fully passing the L<IIIF Image API Validator|https://iiif.io/api/image/validator/>
with all Level 2 plus all optional features expect C<Canonical Link Header>
and some L<inexplicable test failures|https://github.com/nichtich/IIIF/issues/8#issuecomment-545852786>
with PDF, WebP, and JP2 format.

=item

works with ImageMagick 6 (tested on Ubuntu Linux) and ImageMagick 7 (tested on Windows)

=item 

100% L<test coverage|https://coveralls.io/github/nichtich/IIIF> on statement
level, (>90% on branch level and >70% on condition level).

=back

=head1 INSTALLATION

See also L<IIIF::Magick/REQUIREMENTS> for additional installation for optional
features.

=head2 UNIX

Most Unixes include system Perl by default. You should also install ImageMagick and
L<cpanminus|https://metacpan.org/pod/App::cpanminus#INSTALLATION>. For instance at
Ubuntu Linux:

  sudo apt-get install imagemagick cpanminus

To speed up installation of Perl dependencies of this package, optionally:

  sudo apt-get install libplack-perl libplack-middleware-crossorigin-perl

And for optional support of WebP format:

  sudo apt-get install webp libwebp-dev

Then install IIIF with Perl package manager:

  cpanm IIIF

=head2 WINDOWS

Install ImageMagick and Perl, for instance with L<Chocolatey|https://chocolatey.org>:

  choco install imagemagick.tool
  choco install strawberryperl

Then install IIIF with Perl package manager:

  cpanm IIIF

=head1 MODULES

=over

=item L<IIIF::Request>

parse and express an IIIF Image API request build of region, size, rotation, quality, and format

=item L<IIIF::Magick>

get image information and convert images as specified with IIIF Image API request using ImageMagick

=item L<IIIF::ImageAPI>

provide a L<Plack> web service to access images via IIIF Image API

=back

=head1 SCRIPTS

This module provides the command line script L<i3f> to apply IIIF Image API requests without a web service.

=head1 SEE ALSO

=over

=item L<https://github.com/IIIF/awesome-iiif>

=item L<Image::Magick>

=item L<Plack::App::ImageMagick>

=back

=head1 LICENSE

Copyright (C) Jakob Voß.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Jakob Voß E<lt>voss@gbv.deE<gt>

=cut



lib/IIIF/Magick.pm
package IIIF::Magick;
use 5.014001;

our $VERSION = "0.05";

use parent 'Exporter';
our @EXPORT = qw(info available convert convert_command convert_args);

use IPC::Cmd qw(can_run);
use List::Util qw(min);

sub available {
    return can_run("magick") || ( can_run("identify") && can_run("convert") );
}

sub info {
    my $file = shift;

    -f $file or die "$file: No such file\n";
    my $cmd = join ' ', map shell_quote($_), qw(identify -format %Wx%H), $file;

    ( `$cmd` =~ /^(\d+)x(\d+)$/ )
      or die "$file: Failed to get image dimensions";

    return {
        '@context' => 'http://iiif.io/api/image/3/context.json',
        type       => 'ImageService3',
        protocol   => 'http://iiif.io/api/image',
        width      => 1 * $1,
        height     => 1 * $2,
        @_
    };
}

sub convert_args {
    my $req = shift;
    my @args;

    # apply region
    if ( $req->{region} eq 'square' ) {

     # could be simpler in ImageMagick 7:
     # push @args, qw(-gravity center -crop), "%[fx:w>h?h:w]x%[fx:w>h?h:w]+0+0";
        push @args, qw(-set option:distort:viewport
          %[fx:w>h?h:w]x%[fx:w>h?h:w]+%[fx:w>h?(w-h)/2:0]+%[fx:w>h?0:(h-w)/2]
          -filter point -distort SRT 0 +repage);
    }
    elsif ( my $region_px = $req->{region_px} ) {
        my ( $x, $y, $w, $h ) = @$region_px;
        @args = ( '-crop', "${w}x$h+$x+$y" );
    }
    elsif ( my $region_pct = $req->{region_pct} ) {
        my ( $x, $y, $w, $h ) = @$region_pct;

        if ( $x || $y ) {
            my $px = $x / 100;
            my $py = $y / 100;
            push @args, '-set', 'page', "-%[fx:w*$px]-%[fx:h*$py]";
        }

        # could also be simpler in ImageMagick 7
        push @args, '-crop', "${w}x$h%+0+0";
    }

    # apply size
    if ( $req->{size_pct} ) {
        push @args, '-resize', $req->{size_pct} . '%';
    }
    elsif ( $req->{size_px} ) {
        my ( $x, $y ) = @{ $req->{size_px} };

        if ( $x && !$y ) {
            push @args, '-resize', "${x}";
        }
        elsif ( !$x && $y ) {
            push @args, '-resize', "x${y}";
        }
        elsif ( $req->{ratio} ) {
            push @args, '-resize', "${x}x$y";
        }
        else {
            push @args, '-resize', "${x}x$y!";
        }
    }

    # apply rotation
    push @args, '-flop' if $req->{mirror};
    if ( my $degree = $req->{degree} ) {
        push @args, '-rotate', $degree;
        if ( $degree - 90 * int( $degree / 90 ) ) {
            push @args, '-background', 'none';
        }
    }

    # apply quality
    if ( $req->{quality} eq 'gray' ) {
        push @args, qw(-colorspace Gray);
    }
    elsif ( $req->{quality} eq 'bitonal' ) {
        push @args, qw(-monochrome -colors 2);
    }

    return @args;
}

sub convert_command {
    my ( $req, $in, $out ) = splice @_, 0, 3;

    my @cmd = ( 'convert', @_, convert_args($req) );
    push @cmd, $in  if defined $in  and $in ne '';
    push @cmd, $out if defined $out and $out ne '';
    unshift @cmd, "magick" if can_run("magick");

    return join ' ', map shell_quote($_), @cmd;
}

sub convert {
    my $command = convert_command(@_);
    qx{$command};
    return !$?;
}

# adopted from <https://metacpan.org/release/ShellQuote-Any-Tiny>
sub shell_quote {
    my $arg = shift;

    if ( $^O eq 'MSWin32' ) {
        if ( $arg !~ /\A[\w_+-]+\z/ ) {
            $arg =~ s/\\(?=\\*(?:"|$))/\\\\/g;
            $arg =~ s/"/\\"/g;
            return qq("$arg");
        }
    }
    elsif ( $arg !~ qr{\A[\w,_+/.-]+\z} ) {
        $arg =~ s/'/'"'"'/g;
        return "'$arg'";
    }

    return $arg;
}

1;
__END__

=head1 NAME

IIIF::Magick - transform image with IIIF Image API Request using Image Magick

=head1 SYNOPSIS

    use IIIF::Magick qw(info convert);

    my $info = info($file, profile => "level0", id => "...") ;
    
    convert( $request, $file, "target.png" );

=head1 DESCRIPTION

This module maps L<IIIF ImageAPI|https://iiif.io/api/image/3.0/> request
parameters to L<ImageMagick|https://www.imagemagick.org/> command line
arguments. See L<i3f> (command line) and L<IIIF::ImageAPI> (web service)
for applications that make use of it.

=head1 REQUIREMENTS

Function C<info> and C<convert> require ImageMagick to be installed. Converting
to PDF and/or WebP may not be enabled by default. For instance at Ubuntu Linux
remove the line

  <policy domain="coder" rights="none" pattern="PDF" />

fro C</etc/ImageMagick*/policy.xml> and install WebP support via:

  sudo apt-get install webp libwebp-dev

=head1 FUNCTIONS

=head2 available

Returns whether ImageMagick is available.

=head2 info( $file [, id => $id ] [, profile => $profile ] )

Returns L<image information|https://iiif.io/api/image/3.0/#5-image-information>
object with fields C<@context>, C<type>, C<profile>, C<width>, and C<height>.
Fields C<id> and C<profile> must be added for full IIIF compliance.

=head2 convert( $request, $file, $output [, @args ] )

Convert an image file as specified with a L<IIIF::Request> into an output file.
Returns true on success. Additional arguments are prepended to the call of
ImageMagick's C<convert>.

Requires at least ImageMagick 6.9.

=head2 convert_command( $request, $file, $output [, @args ] )

Get a shell-quoted command to convert an image with a L<IIIF::Request>.

=head2 convert_args( $request )

Get the list of command line arguments to C<convert> to transform an image file
as specified via a L<IIIF::Request>.

=head2 LIMITATIONS

The upscale option of L<size|https://iiif.io/api/image/3.0/#42-size> parameter
is ignored: size C<^max> will not upscale the image as the resulting size
depends on additional variables C<maxWidth>, C<maxHeight>, C<maxArea>. 

The IIIF Image API Request is not validated before processing. Sizes larger
than the selected region will therefore always result in an upscaled image.
Use method C<canonical> of L<IIIF::Request> to filter out such invalid
requests.

=cut


lib/IIIF/Request.pm
package IIIF::Request;
use 5.014001;

our $VERSION = "0.06";

use Plack::Util::Accessor qw(region size rotation quality format);
use Carp qw(croak);
use List::Util qw(min);

our $XY  = qr{[0-9]+};               # non-negative integer
our $WH  = qr{[1-9][0-9]*};          # positive integer
our $NUM = qr{[0-9]*(\.[0-9]+)?};    # non-negative
our $REGION   = qr{full|square|($XY,$XY,$WH,$WH)|pct:($NUM,$NUM,$NUM,$NUM)};
our $SIZE     = qr{(\^)?(max|pct:($NUM)|($WH,)|(,$WH)|(!)?($WH,$WH))};
our $ROTATION = qr{([!])?($NUM)};
our $QUALITY  = qr{color|gray|bitonal|default};
our $FORMAT   = qr{[^.]+};

use overload '""' => \&as_string, fallback => 1;

sub new {
    my $class = shift;
    my $path = shift // "";

    my (
        $rotation, $mirror,     $degree,    $quality, $format,
        $region,   $region_pct, $region_px, $size,    $upscale,
        $size_px,  $size_pct,   $ratio
    );

    my @parts = split '/', $path;

    if ( @parts && $parts[0] =~ /^$REGION$/ ) {
        $region = shift @parts;
        if ($1) {
            $region_px = [ split ',', $1 ];
        }
        elsif ($2) {
            $region_pct = [ map { 1 * $_ } split ',', $2 ];
            error("disallowed percentage value")
              if !$region_pct->[2]
              || !$region_pct->[3]
              || grep { $_ > 100 } @$region_pct;
        }
    }

    if ( @parts && $parts[0] =~ /^$SIZE$/ ) {
        $size    = shift @parts;
        $upscale = $1;
        $ratio   = $7;
        $size_px = [ split ',', $5 // $6 // $8 ] if $5 // $6 // $8;

        if ( defined $3 ) {
            $size_pct = 1 * $3;
            if ($upscale) {
                $size = "^pct:$size_pct";
            }
            else {
                error("disallowed percentage value")
                  if $size_pct == 0.0 || $size_pct > 100.0;
                $size = "pct:$size_pct";
            }
        }
    }

    if ( @parts && $parts[0] =~ /^$ROTATION$/ ) {
        shift @parts;
        $mirror = !!$1;

        # normalize to 0...<360 with up to 6 decimal points
        $degree = 1 * sprintf( "%.6f", $2 - int( $2 / 360 ) * 360 );
        $rotation = $mirror ? "!$degree" : "$degree";
    }

    if ( @parts && $parts[0] =~ /^(($QUALITY)([.]($FORMAT))?|[.]($FORMAT))$/ ) {
        $quality = $2;
        $format = $4 // $5;
        shift @parts;
    }

    error( "failed to parse '" . join( '/', '', @parts ) . "'" )
      if @parts;

    bless {
        region => $region // 'full',
        region_pct => $region_pct,
        region_px  => $region_px,
        size       => $size // 'max',
        upscale    => $upscale,
        size_pct   => $size_pct,
        size_px    => $size_px,
        ratio      => $ratio,
        rotation   => $rotation // '0',
        mirror     => $mirror,
        degree     => $degree,
        quality    => $quality // 'default',
        format     => $format
    }, $class;
}

sub error {
    croak "Invalid IIIF Image API Request: $_[0]";
}

sub canonical {
    my ( $self, $width, $height, %max ) = @_;

    # convert region to /full|x,y,w,h/
    my $region = $self->{region};
    if ( $self->{region} eq 'square' ) {
        my $size = min( $width, $height );
        $region = "0,0,$size,$size";
    }
    elsif ( $self->{region_pct} ) {
        my ( $x, $y, $w, $h ) = @{ $self->{region_pct} };
        $x = pct2px( $x, $width );
        $y = pct2px( $y, $height );
        $w = pct2px( $w, $width ) or return;
        $h = pct2px( $h, $height ) or return;
        $region = "$x,$y,$w,$h";
    }
    elsif ( $self->{region_px} ) {    # region outside of image dimensions?
        my ( $x, $y, $w, $h ) = @{ $self->{region_px} };
        return if $x >= $width && $y >= $height;
    }
    $region = 'full' if $region eq "0,0,$width,$height";

    # proceed with region size
    if ( $region ne 'full' ) {
        ( undef, undef, $width, $height ) = split ',', $region;
    }

    if ( $self->{size_pct} ) {        # too small
        return
          if !pct2px( $self->{size_pct}, $width )
          || !pct2px( $self->{size_pct}, $height );
    }

    # convert size to /[^]?(max|w,h)/
    my $size    = $self->{size};
    my $upscale = $self->{upscale};
    my $ratio   = $self->{ratio};
    my $size_px = $self->{size_px};

    my $maxHeight = $max{maxHeight};
    my $maxWidth = $max{maxWidth} || $maxHeight;

    if ( $size eq '^max' && $maxHeight ) {
        $size_px = [ $maxWidth, $maxHeight ];
        $upscale = 1;
        $ratio   = 1;
        $size    = '^!' . join ',', @$size_px;
    }

    if ( $size !~ /\^?max/ ) {
        if ( $self->{size_pct} ) {
            $size = join ',',
              map { pct2px( $self->{size_pct}, $_ ) } ( $width, $height );
            $size = "^$size" if $upscale;
        }
        else {
            my ( $w, $h ) = @$size_px;
            return if !$w && !$h;
            return if !$upscale && ( $h > $height || $w > $width );

            if ( $w && $h ) {
                if ($ratio) {
                    if ( $w / $h > $width / $height ) {
                        $w = pct2px( 100 * $width / $height, $h );
                    }
                    else {
                        $h = pct2px( 100 * $height / $width, $w );
                    }
                }

                $size = "$w,$h";
            }
            elsif ($w) {
                $size = "$w," . pct2px( 100 * $height / $width, $w );
            }
            elsif ($h) {
                $size = pct2px( 100 * $width / $height, $h ) . ",$h";
            }

            $size = "^$size" if $upscale;
        }

        $size = "max" if $size =~ /^\^?$width,$height$/;
    }

    # give up if image too large
    if ($maxHeight) {
        ( $width, $height ) = ( $1, $2 ) if $size =~ /^\^?(\d+),(\d+)$/;
        return if $width > $maxWidth || $height > $maxHeight;
    }

    my $str = join '/', $region, $size, $self->{rotation}, $self->{quality};
    return defined $self->{format} ? "$str.$self->{format}" : $str;
}

sub pct2px {
    my ( $percent, $value ) = @_;
    return int( $percent * $value / 100 + 0.5 );
}

sub is_default {
    my ($self) = @_;

    return $self->as_string =~ qr{^full/max/0/default};
}

sub as_string {
    my ($self) = @_;

    my $str = join '/', map { $self->{$_} } qw(region size rotation quality);
    return defined $self->{format} ? "$str.$self->{format}" : $str;
}

1;
__END__

=head1 NAME

IIIF::Request - IIIF Image API request object

=head1 SYNOPSIS

    use IIIF::Request;

    my $request = IIIF::Request->new('125,15,120,140/90,/!345/gray.jpg');

=head1 DESCRIPTION

Stores parts of an L<IIIF ImageAPI|https://iiif.io/api/image/3.0/> after
C<{identifier}>:

    {region}/{size}/{rotation}/{quality}.{format}

In contrast to the IIIF Image API Specification, all parts are optional.
Omitted parts are set to their default value, except for C<format> which is
allowed to be undefined.  Parsing of percentage and degree values is more
forgiving than required by the specification as values are normalized (e.g.
removal of redundant digits).  The following additional fields are set if
deriveable form the request:

=over

=item region_pct 

=item region_px

=item upscale

=item size_pct

=item size_px

=item ratio

=item mirror

=item degree

=back

=head1 METHODS

=head2 new( [ $request ] )

Parses a request string. It's ok to only include selected image manipulations.
Will raise an error on invalid requests. The request is parsed independent from
a specific image so regions and sizes outside of the image bound are not
detected as invalid.

=head2 as_string

Returns the full request string. Percentage values and degrees are normalized.

=head2 is_default

Returns whether the request (without format) is the default request
C<full/max/0/default> to get an unmodified image.

=head2 canonical( $width, $height [, %max ] )

Returns the L<canonical request|https://iiif.io/api/image/3.0/#47-canonical-uri-syntax>
for an image of given width and height or C<undef> if this would result in an
invalid request (because region or size would be out of bounds). In contrast to
the specification, the C<format> is not required part of the canonical request.

Optional named arguments C<maxWidth> and C<maxHeight> in C<%max> can be used to
control maximum allowed image size. As specified, C<maxWidth> is ignored unless
C<maxHeight> is also given. Option C<maxArea> is not supported.

=head2 error

Raise an "Invalid IIIF Image API Request" error. Can also be used as function.

=cut


lib/IIIF/ImageAPI.pm
package IIIF::ImageAPI;
use 5.014001;

our $VERSION = "0.06";

use parent 'Plack::Component';

use IIIF::Magick qw(info convert);
use File::Spec;
use Try::Tiny;
use Plack::Request;
use IIIF::Request;
use JSON::PP;
use File::Temp qw(tempdir);
use Digest::MD5 qw(md5_hex);
use HTTP::Date;

use Plack::MIME;
Plack::MIME->add_type( '.jp2',  'image/jp2' );
Plack::MIME->add_type( '.webp', 'image/webp' );

use Cwd;
use Plack::Util;

use Plack::Util::Accessor
  qw(images base cache formats rights service canonical magick_args preferredFormats maxWidth maxHeight maxArea);

our @FORMATS = qw(jpg png gif);

sub new {
    my $class = shift;
    my $self  = $class->SUPER::new(@_);

    $self->images('.') unless $self->images;
    $self->formats( [qw{jpg png gif}] ) unless $self->formats;

    $self;
}

sub call {
    my ( $self, $env ) = @_;
    my $req = Plack::Request->new($env);

    if ( $req->path_info =~ qr{^/([^/]+)/?(.*)$} ) {
        my ( $identifier, $request ) = ( $1, $2 );
        if ( my $file = $self->file($identifier) ) {
            $file->{id} = ( $self->base // $req->base ) . $identifier;
            return $self->response( $file, $request );
        }
    }

    return error_response( 404, "Not Found" );
}

sub response {
    my ( $self, $file, $local ) = @_;

    # Image Information Request
    if ( $local eq '' ) {
        return redirect( $file->{id} . "/info.json" );
    }
    elsif ( $local eq 'info.json' ) {
        return $self->info_response($file);
    }

    # allow abbreviated requests, redirect to full form
    my $request = eval { IIIF::Request->new($local) };
    if ($@) {
        return error_response( 400, ( split( " at ", $@ ) )[0] );
    }

    $request->{format} = $request->{format}
      // ( $self->{preferredFormats} || [] )->[0] // $file->{format};

    return error_response( 400, "unsupported format" )
      unless grep { $_ eq $request->{format} } @{ $self->formats };

    if ( !$self->canonical && "$request" ne $local ) {
        return redirect( $file->{id} . "/$request" );
    }

    my $info = info( $file->{path} );
    my $norm = $request->canonical( $info->{width}, $info->{height}, %$self )
      or return error_response();
    $request = IIIF::Request->new($norm);
    my $canonical = $file->{id} . "/$request";

    if ( $self->canonical && "$request" ne $local ) {
        return redirect($canonical);
    }

    # Image Request

    # directly serve unmodified image
    if ( $request->is_default && $request->{format} eq $file->{format} ) {
        return image_response( $file->{path}, $canonical );
    }

    # cache image segment and directly serve if found
    my $cache = $self->cache // $self->cache( tempdir( CLEANUP => 1 ) );
    my $cache_file = File::Spec->catfile( $cache,
        md5_hex("$request") . ".$request->{format}" );

    if ( -r $cache_file ) {
        return image_response( $cache_file, $canonical );
    }
    else {
        my @args = ( $request, $file->{path}, $cache_file );
        push @args, @{ $self->{magick_args} || [] };
        return image_response( $cache_file, $canonical ) if convert(@args);
    }

    error_response( 500, "Conversion failed" );
}

sub info_response {
    my ( $self, $file ) = @_;

    my $info = info(
        $file->{path},
        id      => $file->{id},
        profile => 'level2',

        extraQualities => [qw(color gray bitonal default)],
        extraFormats   => $self->formats,
        extraFeatures  => [
            qw(
              baseUriRedirect canonicalLinkHeader cors jsonldMediaType mirroring
              profileLinkHeader
              regionByPct regionByPx regionSquare rotationArbitrary rotationBy90s
              sizeByConfinedWh sizeByH sizeByPct sizeByW sizeByWh sizeUpscaling
              )
        ]
    );

    $info->{$_} = $self->{$_}
      for grep { $self->{$_} } qw(maxWidth maxHeight maxArea);

    if ( $self->preferredFormats ) {
        $info->{preferredFormats} = $self->preferredFormats;
    }

    # TODO: canonicalLinkHeader?

    $info->{rights}  = $self->rights  if $self->rights;
    $info->{service} = $self->service if $self->service;

    return json_response( 200, $info,
        'application/ld+json;profile="http://iiif.io/api/image/3/context.json"'
    );
}

sub find_file {
    my ( $self, $identifier ) = @_;

    foreach ( @{ $self->formats } ) {
        my $file = File::Spec->catfile( $self->images, "$identifier.$_" );
        return $file if -r $file;
    }
}

sub file {
    my ( $self, $identifier ) = @_;

    my $path =
      ref $self->images eq 'CODE'
      ? $self->images->($identifier)
      : $self->find_file($identifier);

    if ( -f $path && $path =~ /\.([^.]+)$/ ) {
        if ( grep { $1 eq $_ } @{ $self->formats } ) {
            return {
                path   => $path,
                format => $1
            };
        }
    }
}

sub redirect {
    return [ 303, [ Location => $_[0] ], [] ];
}

# adopted from Plack::App::File
sub image_response {
    my ( $file, $canonical ) = @_;

    open my $fh, "<:raw", $file
      or return error_response( 403, " Forbidden " );

    my $type = Plack::MIME->mime_type($file) // 'image';
    my @stat = stat $file;

    Plack::Util::set_io_path( $fh, Cwd::realpath($file) );

    return [
        200,
        [
            'Content-Type'   => $type,
            'Content-Length' => $stat[7],
            'Last-Modified'  => HTTP::Date::time2str( $stat[9] ),
            'Link' => '<http://iiif.io/api/image/3/level2.json>;rel="profile"',
            'Link' => "<$canonical>;rel=\"canonical\""
        ],
        $fh,
    ];
}

sub json_response {
    my ( $code, $body, $type ) = @_;

    state $JSON = JSON::PP->new->pretty->canonical(1);

    [
        $code,
        [
            'Content-Type' => $type // 'application/json',
            'Link' => '<http://iiif.io/api/image/3/level2.json>;rel="profile"'
        ],
        [ $JSON->encode($body) ]
    ];
}

sub error_response {
    my $code = shift // 400;
    my $message = shift
      // " Invalid IIIF Image API Request : region or size out of bounds ";
    json_response( $code, { message => $message } );
}

1;

=head1 NAME

IIIF::ImageAPI - IIIF Image API implementation as Plack application

=head1 SYNOPSIS

    use Plack::Builder;
    use IIIF::ImageAPI;

    builder {
        enable 'CrossOrigin', origins => '*';
        IIIF::ImageAPI->new(
            images  => 'path/to/images',
            base    => 'https://example.org/iiif/',
            formats => [qw(jpg png gif tif pdf webp jp2)],
        );
    }

=head1 CONFIGURATION

=over

=item images

Either an image directory (set to the current directory by default) or a code
reference of a function that maps image identifiers to image files.

=item cache

Cache directory. Set to a temporary per-process directory by default. Please
use different cache directories for different settings of C<maxWidth> and
C<maxHeight>.

=item base

Base URI which the service is hosted at, including trailing slash. Likely
required if the service is put behind a web proxy.

=item canonical

Redirect requests to the L<canonical URI syntax|https://iiif.io/api/image/3.0/#47-canonical-uri-syntax>
and include (disabled by default). A canonical Link header is set anyway.

=item formats

List of supported image formats. Set to C<['jpg', 'png', 'gif']> by default. On
configuration with other formats make sure ImageMagick supports them (see
L<IIIF::Magick/REQUIREMENTS>).

=item preferredFormats

Optional list of preferred image formats. MUST be a subset of or equal to
C<formats>. The first preferred format, if given, will be used as default if a
request does no specify a file format.

=item maxWidth

Optional maximum width in pixels to be supported.

=item maxHeight

Optional maximum height in pixels to be supported.

=item maxArea

Optional maximum pixel area (width x height) to be supported.

=item rights

Optional string that identifies a license or rights statement for all images,
to be included in image information responses.

=item service

Optional array with L<Services|https://iiif.io/api/image/3.0/#58-linking-properties>
to be included in image information responses.

=item magick_args

Additional command line arguments always used when calling ImageMagick. For
instance C<[qw(-limit memory 1GB -limit disk 1GB)]> to limit resources.

=back

=cut
